accepted/future-releases/unified-collections/feature-specification.md

```diff
@@ -0,0 +1,871 @@
+
+    *In other words, inference flows upwards from the body element. Note that
+    both of the above cases can validly apply because of `dynamic` spreads.*
+
+Finally, we define inference on a `setOrMapLiteral` *collection* as follows:
+
+*   If *collection* is syntactically known to be a set literal, then the
+    downwards context for inference of the elements of *collection* is `Set<P>`
+    where `P` may be `?` if downwards inference does not constrain the type of
+    *collection*.
+
+    *   If `P` is `?` then the static type of *collection* is `Set<T>` where `T`
+        is the upper bound of the inferred set element types of the elements.
+
+    *   Otherwise, the static type of *collection* is `Set<T>` where `T` is
+        determined by downwards inference.
+
+    *Note that the element inference will never produce a key/value type here
+    given that downwards context.*
+
+*   If *collection* is syntactically known to be a map literal and downwards
+    inference has statically known type `Map<K, V>` then the downwards context
+    for the elements of *collection* is `Map<K, V>`.
+
+*   If *collection* is syntactically known to be a map literal then the
+    downwards context for the elements of *collection* is `Map<Pk, Pv>` where
+    `Pk` and `Pv` are determined by downwards inference, and may be `?` if the
+    downwards context does not constrain one or both.
+
+    *   If `Pk` is `?` then the static key type of *collection* is `K` where `K`
+        is the upper bound of the inferred key types of the elements.
+
+    *   Otherwise the static key type of *collection* is `K` where `K` is
+        determined by downwards inference.
+
+    And:
+
+    *   If `Pv` is `?` then the static value type of *collection* is `V` where
+        `V` is the upper bound of the inferred value types of the elements.
+
+    *   Otherwise the static value type of *collection* is `V` where `V` is
+        determined by downwards inference.
+
+    *Note that the element inference will never produce a set element type here
+    given this downwards context.*
+
+*   Otherwise, *collection* is not syntactically known to be a set or map
+    literal, then the downwards context for the elements of *collection* is `?`,
+    and the disambiguation is done as follows:
+
+    *   If all elements can be a set, and at least one element must be a set,
+        then *collection* is a set literal with static type `Set<T>` where `T`
+        is the upper bound of the set element types of the elements.
+
+    *   If all elements can be a map, and at least one element must be a map,
+        then *e* is a map literal with static type `Map<K, V>` where `K` is the
+        upper bound of the key types of the elements and `V` is the upper bound
+        of the value types.
+
+    *   If all elements can be both maps and sets, then the literal is ambiguous
+        and it is a compile-time error. *This occurs when all the *leaf
+        elements* are dynamic spreads.*
```

Comment:
occurs, for example, when ...

(because it can also happen for `{...ListMap<T>()}`.